src/Builder/FrontendDialectHelper.cpp

```diff
@@ -11,232 +11,203 @@
-    break;
-  }
-  case (onnx::TensorProto::DOUBLE): {
-    const auto &arrayAttrInitializer =
-        CreateArrayAttribute<double>(initializer);
-    auto elmType = builder.getF64Type();
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::makeArrayRef(arrayAttrInitializer));
-    break;
-  }
-  case (onnx::TensorProto::INT8): {
-    const auto &arrayAttrInitializer =
-        CreateArrayAttribute<int8_t>(initializer);
-    auto elmType = builder.getIntegerType(8);
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::makeArrayRef(arrayAttrInitializer));
-    break;
-  }
-  case (onnx::TensorProto::UINT8): {
-    const auto &arrayAttrInitializer =
-        CreateArrayAttribute<uint8_t>(initializer);
-    auto elmType = builder.getIntegerType(8, false);
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::makeArrayRef(arrayAttrInitializer));
-    break;
-  }
-  case (onnx::TensorProto::INT16): {
-    const auto &arrayAttrInitializer =
-        CreateArrayAttribute<int16_t>(initializer);
-    auto elmType = builder.getIntegerType(16);
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::makeArrayRef(arrayAttrInitializer));
-    break;
-  }
-  case (onnx::TensorProto::INT32): {
-    const auto &arrayAttrInitializer =
-        CreateArrayAttribute<int32_t>(initializer);
-    auto elmType = builder.getIntegerType(32);
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::makeArrayRef(arrayAttrInitializer));
-    break;
-  }
-  case (onnx::TensorProto::INT64): {
-    const auto &arrayAttrInitializer =
-        CreateArrayAttribute<int64_t>(initializer);
-    auto elmType = builder.getIntegerType(64);
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::makeArrayRef(arrayAttrInitializer));
-    break;
-  }
-  case (onnx::TensorProto::BOOL): {
-    const auto &data = CreateArrayAttribute<bool>(initializer);
-    auto elmType = builder.getI1Type();
-    auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
-    denseElmAttr = mlir::DenseElementsAttr::get(
-        tensorType, llvm::SmallVector<bool, 64>(data.begin(), data.end()));
-    break;
-  }
+  llvm::ArrayRef<int64_t> tensorDims(tp.dims().data(), tp.dims().size());
+  mlir::Type elmType = convertONNXTypeToMLIRType(
+      builder, (onnx::TensorProto_DataType)tp.data_type());
+  auto tensorType = mlir::RankedTensorType::get(tensorDims, elmType);
+  auto denseBuilder = [tensorType](auto arrayRef) {
```

Comment:
the use of `auto` is critical here because it's a polymorphic lambda that takes on different types in the different places it is used in the switch statement below